# Segmenting and Clustering Neighborhoods in Toronto City

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [2]:
from bs4 import BeautifulSoup # this module helps in web scrapping.


In [3]:
url= "https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&direction=prev&oldid=926287641"

In [4]:
data = requests.get(url).text

In [5]:
soup= BeautifulSoup(data, "html5lib")

In [6]:
tables= soup.find_all('table')

In [7]:
len(tables)

5

In [8]:
tables

[<table class="wikitable sortable">
 <tbody><tr>
 <th>Postcode</th>
 <th>Borough</th>
 <th>Neighbourhood
 </th></tr>
 <tr>
 <td>M1A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M2A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M3A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
 </td></tr>
 <tr>
 <td>M4A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
 </td></tr>
 <tr>
 <td>M5A</td>
 <td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
 <td><a class="mw-redirect" href="/wiki/Harbourfront_(Toronto)" title="Harbourfront (Toronto)">Harbourfront</a>
 </td></tr>
 <tr>
 <td>M5A</td>
 <td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
 <td><a href="/wiki/Regent_Park" title="Re

In [9]:
for index,table in enumerate(tables):
    if ("Postcode" in str(table)):
        table_index = index
print(table_index)

0


Analyzing from above two code cells, It appears that the table index of the table we need is '0'.

In [10]:
print(tables[table_index].prettify())

<table class="wikitable sortable">
 <tbody>
  <tr>
   <th>
    Postcode
   </th>
   <th>
    Borough
   </th>
   <th>
    Neighbourhood
   </th>
  </tr>
  <tr>
   <td>
    M1A
   </td>
   <td>
    Not assigned
   </td>
   <td>
    Not assigned
   </td>
  </tr>
  <tr>
   <td>
    M2A
   </td>
   <td>
    Not assigned
   </td>
   <td>
    Not assigned
   </td>
  </tr>
  <tr>
   <td>
    M3A
   </td>
   <td>
    <a href="/wiki/North_York" title="North York">
     North York
    </a>
   </td>
   <td>
    <a href="/wiki/Parkwoods" title="Parkwoods">
     Parkwoods
    </a>
   </td>
  </tr>
  <tr>
   <td>
    M4A
   </td>
   <td>
    <a href="/wiki/North_York" title="North York">
     North York
    </a>
   </td>
   <td>
    <a href="/wiki/Victoria_Village" title="Victoria Village">
     Victoria Village
    </a>
   </td>
  </tr>
  <tr>
   <td>
    M5A
   </td>
   <td>
    <a href="/wiki/Downtown_Toronto" title="Downtown Toronto">
     Downtown Toronto
    </a>
   </td>
   <td>
    <a class=

Now, we will convert above table into pandas dataframe.

In [11]:
df = pd.DataFrame(columns=["Postcode", "Borough", "Neighbourhood"])

for row in tables[table_index].tbody.find_all("tr"):
    col = row.find_all("td")
    if (col != []):
        postcode = col[0].text.strip()
        
        borough = col[1].text.strip()
        neighbourhood = col[2].text.strip()
        if borough!= "Not assigned" :
            
            df = df.append({"Postcode":postcode, "Borough":borough, "Neighbourhood":neighbourhood}, ignore_index=True)


# print df

df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M6A        North York   
5        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
9        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
12       M4B         East York   
13       M5B  Downtown Toronto   
14       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M9B         Etobicoke   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M1C       Scarborough   
24       M3C        North York   
25       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
30       M9C         Etobicoke   
31       M9C         Etobicoke   
32       M9C         Etobicoke   
33       M1E       Scarborough   
34       M1E       Scarborough   
35       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
46       M3H        North York   
47       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
50       M5H  Downtown Toronto   
51       M5H  Downtown Toronto   
52       M6H      West Toronto   
53       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
56       M2J        North York   
57       M2J        North York   
58       M3J        North York   
59       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
62       M5J  Downtown Toronto   
63       M5J  Downtown Toronto   
64       M6J      West Toronto   
65       M6J      West Toronto   
66       M1K       Scarborough   
67       M1K       Scarborough   
68       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
71       M3K        North York   
72       M4K      East Toronto   
73       M4K      East Toronto   
74       M5K  Downtown Toronto   
75       M5K  Downtown Toronto   
76       M6K      West Toronto   
77       M6K      West Toronto   
78       M6K      West Toronto   
79       M1L       Scarborough   
80       M1L       Scarborough   
81       M1L       Scarborough   
82       M2L        North York   
83       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
86       M4L      East Toronto   
87       M5L  Downtown Toronto   
88       M5L  Downtown Toronto   
89       M6L        North York   
90       M6L        North York   
91       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
94       M1M       Scarborough   
95       M1M       Scarborough   
96       M2M        North York   
97       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
101      M5M        North York   
102      M6M              York   
103      M6M              York   
104      M6M              York   
105      M6M              York   
106      M9M        North York   
107      M9M        North York   
108      M1N       Scarborough   
109      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
115      M6N              York   
116      M9N              

In [12]:
df.loc[16,'Neighbourhood']= "Cloverdale Mall"
df

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M6A        North York   
5        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
9        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
12       M4B         East York   
13       M5B  Downtown Toronto   
14       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M9B         Etobicoke   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M1C       Scarborough   
24       M3C        North York   
25       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
30       M9C         Etobicoke   
31       M9C         Etobicoke   
32       M9C         Etobicoke   
33       M1E       Scarborough   
34       M1E       Scarborough   
35       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
46       M3H        North York   
47       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
50       M5H  Downtown Toronto   
51       M5H  Downtown Toronto   
52       M6H      West Toronto   
53       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
56       M2J        North York   
57       M2J        North York   
58       M3J        North York   
59       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
62       M5J  Downtown Toronto   
63       M5J  Downtown Toronto   
64       M6J      West Toronto   
65       M6J      West Toronto   
66       M1K       Scarborough   
67       M1K       Scarborough   
68       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
71       M3K        North York   
72       M4K      East Toronto   
73       M4K      East Toronto   
74       M5K  Downtown Toronto   
75       M5K  Downtown Toronto   
76       M6K      West Toronto   
77       M6K      West Toronto   
78       M6K      West Toronto   
79       M1L       Scarborough   
80       M1L       Scarborough   
81       M1L       Scarborough   
82       M2L        North York   
83       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
86       M4L      East Toronto   
87       M5L  Downtown Toronto   
88       M5L  Downtown Toronto   
89       M6L        North York   
90       M6L        North York   
91       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
94       M1M       Scarborough   
95       M1M       Scarborough   
96       M2M        North York   
97       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
101      M5M        North York   
102      M6M              York   
103      M6M              York   
104      M6M              York   
105      M6M              York   
106      M9M        North York   
107      M9M        North York   
108      M1N       Scarborough   
109      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
115      M6N              York   
116      M9N              

More than one neighborhood can exist in one postal code area. So, We will combine rows with same postcode into one row by seperating other neighbourhood name with comma.

In [13]:
drop_row_list=[]
for i in range(len(df)):
    
    for j in range(i+1, len(df)):
        if df.loc[i,'Postcode']==df.loc[j,"Postcode"] and df.loc[i, "Borough"]==df.loc[j, "Borough"]:
            df.loc[i,'Neighbourhood']= df.loc[i,'Neighbourhood']+ ', ' + df.loc[j, "Neighbourhood"]
            drop_row_list.append(j)
drop_row_list= set(drop_row_list)

x=list(drop_row_list)


now, drop the duplicate rows with same postcode.

In [14]:
df.drop(x, axis=0, inplace=True)
df

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
4        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
13       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
21       M1C       Scarborough   
24       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
33       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
52       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
58       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
64       M6J      West Toronto   
66       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
72       M4K      East Toronto   
74       M5K  Downtown Toronto   
76       M6K      West Toronto   
79       M1L       Scarborough   
82       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
87       M5L  Downtown Toronto   
89       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
96       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
102      M6M              York   
106      M9M        North York   
108      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
116      M9N              York   
117      M1P       Scarborough   
120      M2P        North York   
121      M4P   Central Toronto   
122      M5P   Central Toronto   
124      M6P      West Toronto   
126      M9P         Etobicoke   
127      M1R       Scarborough   
129      M2R        North York   
130      M4R   Central Toronto   
131      M5R   Central Toronto   
134      M6R      West Toronto   
136      M7R       Mississauga   
137      M9R         Etobicoke   
141      M1S       Scarborough   
142      M4S   Central Toronto   
143      M5S  Downtown Toronto   
145      M6S      West Toronto   
147      M1T       Scarborough   
150      M4T   Central Toronto   
152      M5T  Downtown Toronto   
155      M1V       Scarborough   
159      M4V   Central Toronto   
164      M5V  Downtown Toronto   
171      M8V         Etobicoke   
174      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
193      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
206      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Harbourfront, Regent Park  
4                     Lawrence Heights, Lawrence Manor  
6                                         Not assigned  
7                                     Islington Avenue  
8                                       Rouge, Malvern  
10         

Now, if the Neigbourhood name is 'Not assigned', we will replace it with its borough name.

In [15]:
for  i,row in df.iterrows():
    if df.loc[i, 'Neighbourhood']=='Not assigned':
        df.loc[i, 'Neighbourhood']= df.loc[i, 'Borough']
df        
    
    
    
            
            

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
4        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
13       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
21       M1C       Scarborough   
24       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
33       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
52       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
58       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
64       M6J      West Toronto   
66       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
72       M4K      East Toronto   
74       M5K  Downtown Toronto   
76       M6K      West Toronto   
79       M1L       Scarborough   
82       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
87       M5L  Downtown Toronto   
89       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
96       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
102      M6M              York   
106      M9M        North York   
108      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
116      M9N              York   
117      M1P       Scarborough   
120      M2P        North York   
121      M4P   Central Toronto   
122      M5P   Central Toronto   
124      M6P      West Toronto   
126      M9P         Etobicoke   
127      M1R       Scarborough   
129      M2R        North York   
130      M4R   Central Toronto   
131      M5R   Central Toronto   
134      M6R      West Toronto   
136      M7R       Mississauga   
137      M9R         Etobicoke   
141      M1S       Scarborough   
142      M4S   Central Toronto   
143      M5S  Downtown Toronto   
145      M6S      West Toronto   
147      M1T       Scarborough   
150      M4T   Central Toronto   
152      M5T  Downtown Toronto   
155      M1V       Scarborough   
159      M4V   Central Toronto   
164      M5V  Downtown Toronto   
171      M8V         Etobicoke   
174      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
193      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
206      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Harbourfront, Regent Park  
4                     Lawrence Heights, Lawrence Manor  
6                                         Queen's Park  
7                                     Islington Avenue  
8                                       Rouge, Malvern  
10         

In [16]:
df.reset_index(drop=True) #reset the index.

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A      Queen's Park   
5        M9A         Etobicoke   
6        M1B       Scarborough   
7        M3B        North York   
8        M4B         East York   
9        M5B  Downtown Toronto   
10       M6B        North York   
11       M9B         Etobicoke   
12       M1C       Scarborough   
13       M3C        North York   
14       M4C         East York   
15       M5C  Downtown Toronto   
16       M6C              York   
17       M9C         Etobicoke   
18       M1E       Scarborough   
19       M4E      East Toronto   
20       M5E  Downtown Toronto   
21       M6E              York   
22       M1G       Scarborough   
23       M4G         East York   
24       M5G  Downtown Toronto   
25       M6G  Downtown Toronto   
26       M1H       Scarborough   
27       M2H        North York   
28       M3H        North York   
29       M4H         East York   
30       M5H  Downtown Toronto   
31       M6H      West Toronto   
32       M1J       Scarborough   
33       M2J        North York   
34       M3J        North York   
35       M4J         East York   
36       M5J  Downtown Toronto   
37       M6J      West Toronto   
38       M1K       Scarborough   
39       M2K        North York   
40       M3K        North York   
41       M4K      East Toronto   
42       M5K  Downtown Toronto   
43       M6K      West Toronto   
44       M1L       Scarborough   
45       M2L        North York   
46       M3L        North York   
47       M4L      East Toronto   
48       M5L  Downtown Toronto   
49       M6L        North York   
50       M9L        North York   
51       M1M       Scarborough   
52       M2M        North York   
53       M3M        North York   
54       M4M      East Toronto   
55       M5M        North York   
56       M6M              York   
57       M9M        North York   
58       M1N       Scarborough   
59       M2N        North York   
60       M3N        North York   
61       M4N   Central Toronto   
62       M5N   Central Toronto   
63       M6N              York   
64       M9N              York   
65       M1P       Scarborough   
66       M2P        North York   
67       M4P   Central Toronto   
68       M5P   Central Toronto   
69       M6P      West Toronto   
70       M9P         Etobicoke   
71       M1R       Scarborough   
72       M2R        North York   
73       M4R   Central Toronto   
74       M5R   Central Toronto   
75       M6R      West Toronto   
76       M7R       Mississauga   
77       M9R         Etobicoke   
78       M1S       Scarborough   
79       M4S   Central Toronto   
80       M5S  Downtown Toronto   
81       M6S      West Toronto   
82       M1T       Scarborough   
83       M4T   Central Toronto   
84       M5T  Downtown Toronto   
85       M1V       Scarborough   
86       M4V   Central Toronto   
87       M5V  Downtown Toronto   
88       M8V         Etobicoke   
89       M9V         Etobicoke   
90       M1W       Scarborough   
91       M4W  Downtown Toronto   
92       M5W  Downtown Toronto   
93       M8W         Etobicoke   
94       M9W         Etobicoke   
95       M1X       Scarborough   
96       M4X  Downtown Toronto   
97       M5X  Downtown Toronto   
98       M8X         Etobicoke   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   
101      M8Y         Etobicoke   
102      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Harbourfront, Regent Park  
3                     Lawrence Heights, Lawrence Manor  
4                                         Queen's Park  
5                                     Islington Avenue  
6                                       Rouge, Malvern  
7          

In [17]:
df.shape

(103, 3)

So, now the dataframe has 103 rows and 3 columns.

In [18]:
df_postcode = df.groupby(['Postcode', 'Borough'])
df['Neighbourhood'].apply(', '.join).reset_index()

df_postcode.head()

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
4        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
13       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
21       M1C       Scarborough   
24       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
33       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
52       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
58       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
64       M6J      West Toronto   
66       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
72       M4K      East Toronto   
74       M5K  Downtown Toronto   
76       M6K      West Toronto   
79       M1L       Scarborough   
82       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
87       M5L  Downtown Toronto   
89       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
96       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
102      M6M              York   
106      M9M        North York   
108      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
116      M9N              York   
117      M1P       Scarborough   
120      M2P        North York   
121      M4P   Central Toronto   
122      M5P   Central Toronto   
124      M6P      West Toronto   
126      M9P         Etobicoke   
127      M1R       Scarborough   
129      M2R        North York   
130      M4R   Central Toronto   
131      M5R   Central Toronto   
134      M6R      West Toronto   
136      M7R       Mississauga   
137      M9R         Etobicoke   
141      M1S       Scarborough   
142      M4S   Central Toronto   
143      M5S  Downtown Toronto   
145      M6S      West Toronto   
147      M1T       Scarborough   
150      M4T   Central Toronto   
152      M5T  Downtown Toronto   
155      M1V       Scarborough   
159      M4V   Central Toronto   
164      M5V  Downtown Toronto   
171      M8V         Etobicoke   
174      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
193      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
206      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Harbourfront, Regent Park  
4                     Lawrence Heights, Lawrence Manor  
6                                         Queen's Park  
7                                     Islington Avenue  
8                                       Rouge, Malvern  
10         

In [19]:
locgeo_df = pd.read_csv('https://cocl.us/Geospatial_data', index_col='Postal Code')



In [20]:
toronto_data = df.join(locgeo_df, on='Postcode')
toronto_data

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
4        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
13       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
21       M1C       Scarborough   
24       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
33       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
52       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
58       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
64       M6J      West Toronto   
66       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
72       M4K      East Toronto   
74       M5K  Downtown Toronto   
76       M6K      West Toronto   
79       M1L       Scarborough   
82       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
87       M5L  Downtown Toronto   
89       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
96       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
102      M6M              York   
106      M9M        North York   
108      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
116      M9N              York   
117      M1P       Scarborough   
120      M2P        North York   
121      M4P   Central Toronto   
122      M5P   Central Toronto   
124      M6P      West Toronto   
126      M9P         Etobicoke   
127      M1R       Scarborough   
129      M2R        North York   
130      M4R   Central Toronto   
131      M5R   Central Toronto   
134      M6R      West Toronto   
136      M7R       Mississauga   
137      M9R         Etobicoke   
141      M1S       Scarborough   
142      M4S   Central Toronto   
143      M5S  Downtown Toronto   
145      M6S      West Toronto   
147      M1T       Scarborough   
150      M4T   Central Toronto   
152      M5T  Downtown Toronto   
155      M1V       Scarborough   
159      M4V   Central Toronto   
164      M5V  Downtown Toronto   
171      M8V         Etobicoke   
174      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
193      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
206      M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Harbourfront, Regent Park  43.654260 -79.360636  
4                     Lawrence Heights, Lawrence Manor  43.718518 -79.464763  
6                                         Queen's Park  43.662301 -79.38